### DISCUSSION TITLE
Ioc and Lucid Models

### DISCUSSION THREAD

**Discussion by mlusca** [adonisjs/core#1790] [Help] [answered]
I have a singleton service and I'm trying to query a model on it but I encounter the following error when compiling:

`E_IOC_LOOKUP_FAILED: Cannot resolve "Adonis/Lucid/Orm" namespace from the IoC Container`

I need to work with lucid in the container, does anyone know how?

**Comment by Banashek** [comment]
It looks like your service is attempting to be created before the ORM is loaded into the IoC container.

Could you paste your provider class in here?

An example provider that depends on other services can be found in [the Auth package's provider](https://github.com/adonisjs/auth/blob/develop/providers/AuthProvider.ts).

Is your provider specifying the `Adonis/Lucid/Orm` package in it's `boot` method using `this.application.container.with(['Adonis/Lucid/Orm'])` similar to how the Auth provider does for `'Adonis/Core/HttpContext'`?

**Reply by mlusca** [reply]
Provider:
```

import {IocContract} from '@adonisjs/fold'
import AuthenticateService from "App/Modules/Tecnologia/Services/AuthenticateService";

export default class AppProvider {
  constructor(protected $container: IocContract) {
  }

  public register() {
    this.$container.singleton('Mota/Auth', function () {
      return new AuthenticateService()
    })
  }

  public boot() {
    // IoC container is ready
    this.$container.with(['Adonis/Lucid/Orm', 'Mota/Auth'], (Auth) => {  
    })
  }
}
```



Service:
```
import AuthenticateInterface from "Contracts/interfaces/authenticate.interface";
import UsuariosToken from "App/Models/UsuariosToken";

export default class AuthenticateService implements AuthenticateInterface {
  /**
   * Constructor.
   */
  constructor() {

  }

  public tratarStringToken(bearerToken: string): string {
    const token = bearerToken.split('Bearer ')

    return token[1]
  }

  public retornarUsuarioLogado(token: string) {
    console.log(UsuariosToken)
    // return UsuariosToken.query()
    //   .where('usuarios_tokens.token', token)
    //   .join('usuarios as u', 'u.id', 'token.usuario_id')
  }
}
```


What am I doing wrong?

**Comment by mlusca** [comment, marked as answer]
I managed to do so in the register method:

```
this.$container.singleton('Mota/Auth', () => {
      const data = this.$container.use('Adonis/Lucid/Database')

      return new AuthenticateService(data)
    })
```

But I really wanted to use the models

**Reply by Banashek** [reply]
This seems closer. The service needs its dependencies injected into it (either via explicit constructor parameter or via a IoC `use` call).

What if you instead do something like:

```
this.$container.singleton('Mota/Auth', () => {
      const User = this.$container.use('App/Models/User');

      return new AuthenticateService(User);
    });
```

It would also mean just depending on the specific models instead of depending on the entire database.

Not sure if it works but I think that somewhere you need to be more explicit about the dependencies of your provider/service.
